---
title: "South African RSTR Data cleaning"
author: "Kim Dill-McFarland, kadm@uw.edu"
output:
  html_document:
    toc: yes
    toc_float:
      collapsed: no
date: "version `r format(Sys.time(), '%B %d, %Y')`"
editor_options: 
  chunk_output_type: console
---
# Background

DESCRIPTION

The purpose of this workflow is to complete basic data cleaning of metadata and RNA-seq libraries generated from the above experiments. This includes 1) removing low coverage libraries, 2) filtering rare genes, 3) removing outlying libraries, and 4) normalizing for RNA composition. 

# Setup
Load packages

```{r setup, message=FALSE, warning=FALSE}
# Data manipulation and figures
library(tidyverse)
library(janitor)
  # Modify ggplot figures to non-overlapping text labels
  library(ggrepel)
  # Modify ggplot data order within facets
  library(drlib)
  # Plot log scales
  library(scales)
  #Multi-panel figures
  library(cowplot)

# Empirical analysis of digital gene expression data
## Data normalization
library(edgeR)

# Print pretty table to knit file
library(knitr)
library(kableExtra)
  options(knitr.kable.NA = '')
  
#Create 'not in' operator
`%notin%` <- Negate(`%in%`)
```

Set seed

```{r}
set.seed(4389)
```

# Customization

Set variable names and cutoffs for this workflow.

```{r custom}
#Provide file basename
#This will be appended to the start of all file names saved to the disk
basename <- "RSTR_SA"

#List main variables of interest
#Will be used to visualize data in tables and plots
#If > 2 variables provided, plots are not colored by variables
vars_of_interest <- c("condition","Sample_Group",
                      "gender","age","bmi","ethnic",
                      "occupationlevel","workingunderground",
                      "YearsWorkedUnderground",
                      "bcgscar")

# Define cutoffs
#Total sequences MINIMUM
count.cut <- 1E6
```

# Read in and format data 
## Counts and metadata

```{r}
#Raw counts table
counts <- readRDS("data_raw/Hawn_SAMacrophage_counts.rds") %>% 
  as.data.frame() %>% 
  rownames_to_column("geneName")
```

```{r message=FALSE}
#Get donors in RNAseq dataset
meta <- counts[1,] %>% 
  pivot_longer(-geneName, names_to = "libID") %>% 
  select(libID) %>% 
  separate(libID, into=c("FULLIDNO","condition"), sep="_", remove = FALSE) %>% 
  #add addtl metdata
  left_join(read_csv("data_raw/2020.11.05SA.RSTR_Hawn_metadata.csv"),
            by = c("FULLIDNO"="ptID")) %>% 
  #Fill in missing RSTR
  mutate(Sample_Group = ifelse(is.na(Sample_Group) & status == "control",
                               "LTBI",
                               ifelse(is.na(Sample_Group) & status == "case",
                               "RSTR", Sample_Group))) %>% 
  arrange(libID)
```

#### Check samples

*all libraries in the count data are found in the sample metadata. And vice versa.*

```{r echo=FALSE}
identical(colnames(counts)[-1], meta$libID)
```

## Summarize samples

```{r echo=FALSE}
meta %>% 
  count(Sample_Group, condition) %>% 

kable(align="l", caption="Total libraries") %>% 
  kable_styling(bootstrap_options = "striped", full_width = FALSE)
```

# Data cleaning
## Remove select samples

Remove white participants (6 RSTR) and those that converted/reverted TST or IGRA at 12 months (3 RSTR, 1 LTBI).

```{r}
meta.filter <- meta %>% 
  #Rm white participants
  filter(ethnic != "White/European") %>%
  #Rm converters
  filter(FULLIDNO %notin% c("890-01347-2", "890-01612-7", 
                        "890-00658-6", "890-00798-1"))

counts.filter <- counts[,c("geneName", meta.filter$libID)]
```

## Assess library coverage

Calculate total sequences.

```{r}
tot.seq <- counts.filter %>% 
  pivot_longer(-geneName, names_to = "libID") %>% 
  group_by(libID) %>% 
  summarise(tot.seq = sum(value, na.rm=TRUE), .groups="drop")

#add to metadata
meta.filter <- meta.filter %>% 
  full_join(tot.seq)
```

Plot total sequences per library. The minimum total sequences cutoff set above is indicated by a horizontal line.

```{r tot.seqs, echo=FALSE, fig.width=8.5}
meta.filter %>% 
    ggplot(aes(x=reorder_within(FULLIDNO, by=tot.seq, 
                            within=condition), 
           y=tot.seq))  +
    geom_col() +
    # Facet by variable of interest
    facet_wrap(Sample_Group~condition, scales="free_x") +
  # Add cutoff line
  geom_hline(yintercept = count.cut)+
  # Beautify
  theme_classic() +
  theme(axis.text.x = element_text(size=rel(0.75),angle = 90, hjust = 1),
        legend.position = "bottom") +
  labs(x="Library", y="Total aligned counts\n(Log scale)", fill="") +
  scale_x_reordered() +
  scale_y_continuous(trans = 'log10',
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x)))
```

## Filter by library coverage

All libraries have sufficient coverage. Thus, none will be removed.

## Filter protein coding genes

Load key.

```{r key, message=FALSE}
key <- read_tsv("data_raw/2020.06.18_HGNC.gene.key.txt") %>% 
  clean_names() %>% 
  #get all prev and alias symbols
  mutate(previous_symbols = strsplit(as.character(previous_symbols), ", ")) %>% 
  mutate(alias_symbols = strsplit(as.character(alias_symbols), ", ")) %>% 
  unnest(cols=previous_symbols) %>% 
  unnest(cols=alias_symbols) %>% 
  rownames_to_column() %>% 
  pivot_longer(c(approved_symbol:alias_symbols), values_to = "geneName") %>% 
  drop_na(geneName) %>% 
  select(-rowname)
```

Filter gene key to protein coding (pc) genes that occur in the count data set and have valid hgnc symbols. 

```{r message=FALSE}
key.pc <- key %>% 
  # Keep only valid hgnc symbols
  drop_na(geneName) %>% 
  # Keep protein coding genes only
  filter(locus_group == "protein-coding gene") %>% 
  # Remove duplicate entries
  distinct(geneName, .keep_all=TRUE) %>% 
  # Keep only genes found in dataset 
  filter(geneName %in% counts.filter$geneName) %>% 
  arrange(geneName)
```

Filter the count data to pc genes as well.

```{r message=FALSE}
counts.filter.pc <- counts.filter %>% 
  filter(geneName %in% key.pc$geneName) %>% 
  arrange(geneName)
```

#### Check genes

*all genes in the key are found in the data. And vice versa.*

```{r echo=FALSE}
identical(key.pc$geneName, counts.filter.pc$geneName)
```

## Assess PCA
### Variables of interest

If one or more variables of interest represents significant variation in the data, consider further filtering and normalizing separately. This is most common when multiple sample or cell types exist in the data set. 

```{r PCA.all, echo=FALSE, warning=FALSE, fig.height=20}
#Calculate PCA for all data.
PCA.all <- counts.filter.pc %>% 
  column_to_rownames("geneName") %>% 
  #Convert to log counts per million
  cpm(., log=TRUE) %>% 
  t() %>% 
  #Calc PCA
  prcomp()

PC1.label <- paste("PC1 (", summary(PCA.all)$importance[2,1]*100, "%)", sep="")
PC2.label <-paste("PC2 (", summary(PCA.all)$importance[2,2]*100, "%)", sep="")

# Extract PC values
PCA.all.dat <- as.data.frame(PCA.all$x) %>% 
  rownames_to_column("libID") %>%
  # Select PCs for plotting
  dplyr::select(libID, PC1:PC3) %>% 
  # Merge with metadata
  full_join(meta.filter, by="libID")

PCA.list <- list()
#Loop through variables of interest
for(i in 1:length(vars_of_interest)){
  PCA <- ggplot(PCA.all.dat,
                  aes(PC1, PC2)) +
           geom_point(aes_string(color=vars_of_interest[i]),
                      size=3) + 
  #Beautify
  theme_classic() +
  theme(legend.position = "bottom") +
  labs(x=PC1.label, y=PC2.label, title="Un-normalized logCPM") +
  coord_fixed(ratio=1)
  
  PCA.list[[i]] <- PCA
}

plot_grid(plotlist=PCA.list, ncol=2)
```

If you decide to split the data set, continue from [Split data filtering](#split.data).

### PCA outliers

Visualize PCA outliers defined as any library with PC1 and/or PC2 values greater than 3 standard deviations away from the PC mean.

```{r PCA.outlier, echo=FALSE}
PCA.all.sd <- PCA.all.dat %>% 
  group_by(condition) %>% 
  #Calculate PC mean std deviation
  summarize(
    PC1.mean = mean(PC1),
    PC1.sd = sd(PC1),
    PC2.mean = mean(PC2),
    PC2.sd = sd(PC2)) %>% 
  #Calculate +/- 3 sd limits
  summarize(
    PC1.min = PC1.mean-(3*PC1.sd),
    PC1.max = PC1.mean+(3*PC1.sd),
    PC2.min = PC2.mean-(3*PC2.sd),
    PC2.max = PC2.mean+(3*PC2.sd))
  
PCA.all.dat <- PCA.all.dat %>%   
  #ID potential outliers
  mutate(col.group = ifelse(PC1 > PCA.all.sd$PC1.max | 
                            PC1 < PCA.all.sd$PC1.min |
                            PC2 > PCA.all.sd$PC2.max | 
                            PC2 < PCA.all.sd$PC2.min, 
                            "potential outlier", "okay"))
PCA.all.dat %>%       
  ggplot(aes(PC1, PC2)) +
  geom_point(aes(color=col.group),size=3) + 
  geom_text_repel(data=filter(PCA.all.dat,
                            col.group == "potential outlier"),
            aes(label=libID), show.legend = FALSE) +
  #Beautify
  theme_classic() +
  theme(legend.title = element_blank(),
        legend.position = "bottom") +
  labs(x=PC1.label, y=PC2.label, 
       title="Un-normalized logCPM") +
  coord_fixed(ratio=1) +
  scale_color_manual(values = c("#969696","#b10026"))
```

## Filter PCA outliers

No PCA outliers are apparent. Thus, no libraries are removed.

## Create DGEList object

For use in gene filtering with edgeR

```{r dgelist}
dat <- DGEList(
  #count table. move gene names to column names
  counts=as.matrix(column_to_rownames(counts.filter.pc,
                                      "geneName")),
  #metadata
  samples=meta.filter,
  #keep genes in count table
  genes=key.pc)
```

## Filter rare genes

```{r echo=FALSE}
min.CPM <- 0.5
min.pct <- 5

model.mat <- model.matrix(~ condition*Sample_Group, data=dat$samples)
```

The raw gene sets contain highly variable, low abundance/rare genes. Filter genes without > `r min.CPM` CPM in at least `r min.pct`% (`r round(nrow(dat$samples)*min.pct/100)`) samples.

```{r voom1, echo=FALSE, warning=FALSE}
temp <- voom(dat, design=model.mat,
             plot=FALSE, save.plot = TRUE)

MV.plot1 <- data.frame(
  x = temp$voom.xy$x, 
  y = temp$voom.xy$y,
  linex = temp$voom.line$x, 
  liney = temp$voom.line$y) %>% 
  
  ggplot() +
  geom_point(aes(x=x, y=y), size=0.5) +
  geom_path(aes(x=linex, y=liney), color="red") +
  theme_classic() +
  labs(x="log2( count size + 0.5 )", y="Sqrt (stdev)",
       title="All genes")
```

```{r}
source("https://raw.githubusercontent.com/kdillmcfarland/R_bioinformatic_scripts/master/RNAseq_rare_gene_filter.R")

rare.gene.filter(dat, 
                 min.CPM=min.CPM,
                 min.pct=min.pct, 
                 name="dat.abund")
```

```{r voom2, echo=FALSE}
#Assess
temp2 <- voom(dat.abund, design=model.mat,
     plot=FALSE, save.plot = TRUE)

#plot mean-var trend
MV.plot2 <- data.frame(
  x = temp2$voom.xy$x, 
  y = temp2$voom.xy$y,
  linex = temp2$voom.line$x, 
  liney = temp2$voom.line$y) %>% 
  
  ggplot() +
  geom_point(aes(x=x, y=y), size=0.5) +
  geom_path(aes(x=linex, y=liney), color="red") +
  theme_classic() +
  labs(x="log2( count size + 0.5 )", y="Sqrt (stdev)",
       title="Abundant genes")

plot_grid(MV.plot1, MV.plot2)
```

```{r echo=FALSE}
#Count genes removed
genes <- nrow(dat$genes)
genes.abund <- nrow(dat.abund$genes)
```

This removes `r genes-genes.abund` (`r round((genes-genes.abund)/genes*100, digits=0)`%) genes. 

## Normalize for RNA composition

Calculate factors to scale library sizes.

```{r norm}
dat.abund.norm <- calcNormFactors(dat.abund)
```

## Normalize with voom

```{r voom3}
dat.abund.norm.voom <- voomWithQualityWeights(
  dat.abund.norm, design=model.mat, plot=TRUE)
```

## PCA final data set

```{r PCA.final, echo=FALSE, warning=FALSE, fig.height=20}
#Calculate PCA for voom norm data.
PCA.voom <- as.data.frame(dat.abund.norm.voom$E) %>% 
  t() %>% 
  #Calc PCA
  prcomp()

PC1.label <- paste("PC1 (", summary(PCA.voom)$importance[2,1]*100, "%)", sep="")
PC2.label <-paste("PC2 (", summary(PCA.voom)$importance[2,2]*100, "%)", sep="")

# Extract PC values
PCA.voom.dat <- as.data.frame(PCA.voom$x) %>% 
  rownames_to_column("libID") %>%
  # Select PCs for plotting
  dplyr::select(libID, PC1:PC3) %>% 
  # Merge with metadata
  full_join(as.data.frame(dat.abund.norm.voom$targets),
            by="libID")

PCA.list <- list()
#Loop through variables of interest
for(i in 1:length(vars_of_interest)){
  PCA <- ggplot(PCA.voom.dat,
                  aes(PC1, PC2)) +
           geom_point(aes_string(color=vars_of_interest[i]),
                      size=3) + 
  #Beautify
  theme_classic()  +
  theme(legend.position = "bottom") +
  labs(x=PC1.label, y=PC2.label, title="Voom normalized log2 CPM") +
  coord_fixed(ratio=1)
  
  PCA.list[[i]] <- PCA
}

plot_grid(plotlist=PCA.list, ncol=2)
```

## Summarize cleaning

```{r echo=FALSE}
meta %>%
  #ID filters
  mutate(
    white = ifelse(ethnic == "White/European", 
                              "White/European", NA),
    convert = ifelse(FULLIDNO %in% c("890-01347-2", "890-01612-7", 
                        "890-00658-6", "890-00798-1"),
                        "convert/revert", NA)) %>% 
  #Combine filter tags
  group_by(libID) %>% 
  mutate(filter = paste(white,convert, sep=", ")) %>% 
  mutate(filter = gsub("NA, ","",filter)) %>% 
  mutate(filter = gsub(", NA","",filter)) %>% 
  mutate(filter = gsub("NA","",filter)) %>% 
  select(libID, condition, Sample_Group, filter) %>% 
  #Keep data on removed samples only
  filter(filter != "") %>% 
  arrange(libID) %>% 
kable(align="l", 
        caption="Libraries removed") %>% 
  kable_styling(bootstrap_options = "striped", full_width = FALSE)
```

No libraries were removed for quality or PCA. This leaves the following for analysis.

```{r echo=FALSE}
meta.filter %>% 
  count(Sample_Group, condition) %>% 

kable(align="l", caption="Total libraries") %>% 
  kable_styling(bootstrap_options = "striped", full_width = FALSE)
```

# Save data

Write as RData

```{r}
dir.create("data_clean/", showWarnings = FALSE)
save(dat.abund.norm.voom,
     file=paste("data_clean/", basename, "_dat_clean.RData", sep=""))
```

Write counts as table.

```{r}
#Counts table
write_csv(as.data.frame(dat.abund.norm.voom$E),
          paste("data_clean/", basename, "_counts_voom.csv", sep=""))
```

# R session

```{r}
sessionInfo()
```

***